target/linux/ath79/dts/qca9558_belkin_f9j1108-v2.dts

```diff
@@ -0,0 +1,228 @@
+			gpios = <&gpio 19 GPIO_ACTIVE_LOW>;
+		};
+
+		led_system: system {
+			label = "f9j1108-v2:blue:system";
+			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
+		};
+	};
+	
+	gpio-export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+
+		gpio_usb_power {
+			gpio-export,name = "f9j1108-v2:usb2:power";
+			gpio-export,output = <1>;
+			gpios = <&gpio 21 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		wps {
+			label = "wps";
+			linux,code = <KEY_WPS_BUTTON>;
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+		
+		reset {
+			label = "Reset button";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+	
+	virtual_flash {
+		compatible = "mtd-concat";
+
+		devices = <&fwpart1 &fwpart2>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				compatible = "openwrt,okli";
+				label = "firmware";
+				reg = <0x0 0x0>;
+			};
+		};
+	};
+};
+
+&pcie1 {
+	status = "okay";
+};
+
+&uart {
+	status = "okay";
+};
+
+&usb_phy0 {
+	status = "okay";
+};
+
+&usb0 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	status = "okay";
+
+	hub_port0: port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
+};
+
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "winbond,w25q128", "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <80000000>;
+		m25p,fast-read;
+
+		mtdparts: partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			
+			uboot: partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x040000>;
+				read-only;
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
+				reg = <0x040000 0x010000>;
+			};
+			
+			fwpart1: partition@50000 {
+				label = "fwpart1";
+				reg = <0x050000 0xe20000>;
+			};
+			partition@e70000 {
+				label = "loader";
+				reg = <0xe70000 0x10000>;
+				read-only;
+			};
+			fwpart2: partition@e80000 {
+				label = "fwpart2";
+				reg = <0xe80000 0x150000>;
+			};
+			
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x010000>;
+			};
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+
+		qca,ar8327-initvals = <
+			0x04 0x00080080 /* PORT0 PAD MODE CTRL */
+			0x0c 0x07600000 /* PORT6 PAD MODE CTRL */
+			0x50 0xc737c737 /* LED_CTRL0 */
+			0x54 0x00000000 /* LED_CTRL1 */
+			0x58 0x00000000 /* LED_CTRL2 */
+			0x5c 0x0030c300 /* LED_CTRL3 */
+			0x7c 0x0000007e /* PORT0_STATUS */
+			0x94 0x0000007e /* PORT6 STATUS */
+			>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	mtd-mac-address = <&art 0x0>;
+	mtd-mac-address-increment = <(-1)>;
+	phy-handle = <&phy0>;
+	pll-data = <0xa6000000 0x00000101 0x00001616>;
+
+	gmac-config {
+		device = <&gmac>;
+		rgmii-enabled = <1>;
+	};
+};
+
+&eth1 {
+	status = "okay";
+
+	mtd-mac-address = <&art 0x0>;
+	mtd-mac-address-increment = <(0)>;
```

Comment:
Increment 0 makes no sense and can be dropped. What's the vendor MAC address assignment (last byte for lan/wan/wifi) and in which locations do you find addresses (art 0x0, 0x6, 0x1002, 0x5006?)

Reply:
art 0x0 contains eth1 (WAN, port 1 on the switch), and art 0x6 is empty or can contain the eth0 (LAN, port 2-5 on the switch) only after using progmac2 in u-boot.
They use the same MAC for WLAN and LAN, which is -01 compared to art 0x0, or also present at 0x1002.  There is nothing present at 0x5006 (null).

Reply:
So, we can not rely on art 0x6 to be always present. Thus, using 0x0 and 0x0 with offset -1 seems appropriate for ethX. Please make sure the assignment is correct on the device running OpenWrt, as for some devices &eth0/&eth1 in DTS are swapped compared to eth0/eth1 on the running device.

If a MAC address is (always) present in 0x1002, you can drop the mtd-mac-address for wmac, is this location will automatically be read by `mtd-cal-data = <&art 0x1000>;`.

Is there a label on the device and which addresses are shown there?

Reply:
On the label WLAN and LAN both are art 0x0 <(-1).  WAN is exactly as per label at art 0x0.
WLAN (2.4ghz) MAC address should "always" be present in 0x1002, but I'd have to buy another test device to make sure, maybe someone with the EU or US model can verify ? 
WLAN (5ghz) MAC is not on the label, but factory firmware uses whatever is contained within the oversized mini-PCIe card.

Reply:
my EU variant has on the label:
```
LAN/WLAN MAC: xxxxxxxxxx5A     Serial#: 1xxxxxxxxxxxxxx
WAN MAC: xxxxxxxxxx5B                PIN code: xxxxxxxx
                                                                                       Ver.2001az
```

inside the ART partition there is:
- WAN MAC at 0x0
- right after the WAN MAC there's the LAN MAC
- 2.4GHz WLAN MAC at 0x1002 (same as the LAN MAC)
- I cannot find the 5GHz WLAN MAC anywhere inside the ART partition (as I said elsewhere: I believe that it's stored inside an actual EEPROM on the 5GHz WLAN mini-PCIe card)

Reply:
Okay, so the MAC assignment seems correct. For the (first) label MAC, just add the following to the aliases block:
`label-mac-device = &eth0;`

Please also add the information we gained to the commit message; something like that:
```
MAC addresses based on OEM firmware:

interface     address     location
lan              *:5a           sometimes in 0x6
wan            *:5b           0x0
2.4g            *:5b           0x1002
5g               ???            on dedicated EEPROM?
```

Reply:
personally I would omit the colons and upper-case all characters in the MAC addresses.
that's how the MAC addresses are printed on the label of my ``F9K1115V2`` (also printed on the label above all of what's shown above)